Fix for MGXS subdomain averaging

wbinventor opened the pull request:
These tiny PR fixes an issue in the `openmc.mgxs.MGXS.get_subdomain_avg_xs(...)` routine. Presently, this routine computes a simple geometric average of the MGXS across multiple subdomains (*e.g.*, distribcell instances). Now, the average is scalar flux-weighted. This is accurate approach which can be confirmed if one creates a geometry with a distinct "cell" for the particular distribcell instances one wishes to average across. With this fix, the MGXS tallied in the distinct cell will match precisely those computed with scalar flux-weighted subdomain averaging of the corresponding distribcell instances in the original geometry.

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -824,9 +824,9 @@ def get_subdomain_avg_xs(self, subdomains='all'):
-            std_dev = np.mean(std_dev[subdomains, ...]**2, axis=0)
+            # Get the mean, std. dev. across requested subdomains
+            mean = np.sum(mean[subdomains, ...], axis=0)
+            std_dev = np.sum(std_dev[subdomains, ...]**2, axis=0)
```
I don't quite understand how this change reflects what you described in words. Can you elaborate?

wbinventor replied:
The earlier version computed the geometric average of the flux and reaction
rate in each energy group and subdomain. Then it used these geometric
averages to compute the subdomain-averaged MGXS in each energy group.

The new version sums the flux and reaction rate tallies across subdomains
to compute the aggregate. This is effectively the flux weighted average of
the MGXS in each energy group across subdomains. One way to think about
this is that those subdomains with little flux are weighted less in the
average than those subdomains with more flux. This is exactly what is done
for a cell-based tally where the cell is replicated throughout a geometry.
On Nov 9, 2015 9:02 AM, "Paul Romano" <notifications@github.com> wrote:

> In openmc/mgxs/mgxs.py
> <https://github.com/mit-crpg/openmc/pull/496#discussion_r44276963>:
>
> > @@ -824,9 +824,9 @@ def get_subdomain_avg_xs(self, subdomains='all'):
> >              mean = tally.get_reshaped_data(value='mean')
> >              std_dev = tally.get_reshaped_data(value='std_dev')
> >
> > -            # Get the mean of the mean, std. dev. across requested subdomains
> > -            mean = np.mean(mean[subdomains, ...], axis=0)
> > -            std_dev = np.mean(std_dev[subdomains, ...]**2, axis=0)
> > +            # Get the mean, std. dev. across requested subdomains
> > +            mean = np.sum(mean[subdomains, ...], axis=0)
> > +            std_dev = np.sum(std_dev[subdomains, ...]**2, axis=0)
>
> I don't quite understand how this change reflects what you described in
> words. Can you elaborate?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/496/files#r44276963>.
>

paulromano replied:
Maybe my brain is not working because it's Monday morning.. isn't the only difference between `np.mean` and `np.sum` a factor of *N* being the length of the array?

wbinventor commented:
The issue isn't your brain, it is mine - I forgot why I implemented this in the first place late last night and definitely misled you with the description of this PR. The MGXS computed with the new and old versions will be identical (yes, that's right). 

The reason I am using `np.sum` rather than `np.mean` now is that the I want the underlying base tallies in the subdomain-avg MGXS to mirror what one would compute if they were to spend the effort to create distinct cells for each distribcell instance of interest and tally MGXS for each one individually (which is order(s) of magnitude slower than using distribcells btw).

Not to pull OpenMOC into this discussion, but the real reason for doing this is related to the new [SPH factor calculator](https://github.com/mit-crpg/OpenMOC/pull/219) in OpenMOC. In the SPH scheme, we compute a fixed neutron source from OpenMC's fission + scatter reaction rates in each heterogeneous spatial zone from the MGXS and flux. The current implementation assumes that if the base tallies for an MGXS (*e.g.*, flux and reaction rates) are summed across all instances of a cell / material / universe just as they are for a "cell," "material" or "universe" tally in OpenMC. In order to make subdomain-averaged MGXS plug-n-play into the SPH factor calculator, I wanted the base tallies in the MGXS to be consistent with the what one would obtain from an explicitly repreprented "cell" tally.

Again, sorry for the confusion here, I was going all over the place when I implemented this and mis-led myself as to why I needed this before I submitted the PR.

paulromano commented:
Ha, ok. That makes at least a little more sense to me!
